File: src/scripts/lint_positioning.py
#!/usr/bin/env python3
"""Positioning consistency lint for event4u/agent-config.

Asserts that three public-positioning surfaces agree on the canonical
phrasing and that every advertised GitHub topic is discoverable in the
README body (literally or through `equivalents:` paraphrases).

Sources:
  - README.md              — canonical phrasing (H1 + first blockquote)
  - package.json           — `description` field
  - .github/about.yml      — `description` field
  - .github/topics.yml     — `topics:` + optional `equivalents:` map

Failure mode is a diff, not a stack trace. See R5 Phase 3 in
agents/roadmaps/strategic-visibility-mcp-topics-positioning.md.
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

try:
    import yaml
except ImportError:
    sys.stderr.write("❌  PyYAML is required: pip3 install pyyaml\n")
    sys.exit(2)

REPO_ROOT = Path(__file__).resolve().parent.parent.parent
README = REPO_ROOT / "README.md"
PACKAGE_JSON = REPO_ROOT / "package.json"
ABOUT_YML = REPO_ROOT / ".github" / "about.yml"
TOPICS_YML = REPO_ROOT / ".github" / "topics.yml"

# Canonical anchor: the right-side phrase of the README H1
# ("Agent Config — Universal AI Agent OS" → "Universal AI Agent OS").
# This is the substring that MUST appear in the other two surfaces.
H1_RE = re.compile(r"^#\s+(.+?)\s*$", re.MULTILINE)
BLOCKQUOTE_RE = re.compile(r"^>\s+(.+?)\s*$", re.MULTILINE)
DESCRIPTION_MAX = 200


def _read_readme_anchors() -> tuple[str, str, str]:
    text = README.read_text(encoding="utf-8")
    h1_match = H1_RE.search(text)
    if not h1_match:
        raise SystemExit("❌  README.md has no H1 heading")
    h1 = h1_match.group(1)
    # Anchor = the phrase after the em dash, falling back to the whole H1.
    parts = re.split(r"\s+[—–-]\s+", h1, maxsplit=1)
    anchor = parts[1].strip() if len(parts) == 2 else h1.strip()

    bq_match = BLOCKQUOTE_RE.search(text)
    blockquote = bq_match.group(1).strip() if bq_match else ""
    return h1, anchor, blockquote


def _read_package_description() -> str:
    data = json.loads(PACKAGE_JSON.read_text(encoding="utf-8"))
    return str(data.get("description", "")).strip()


def _read_about_description() -> str:
    data = yaml.safe_load(ABOUT_YML.read_text(encoding="utf-8")) or {}
    return str(data.get("description", "")).strip()


def _read_topics() -> tuple[list[str], dict[str, list[str]]]:
    data = yaml.safe_load(TOPICS_YML.read_text(encoding="utf-8")) or {}
    topics = [str(t) for t in data.get("topics", [])]
    equivalents = {str(k): [str(v) for v in vs] for k, vs in (data.get("equivalents") or {}).items()}
    return topics, equivalents


def _topic_present(readme_lc: str, topic: str, equivalents: dict[str, list[str]]) -> tuple[bool, str | None]:
    needles = [topic, topic.replace("-", " "), topic.replace("-", "")] + equivalents.get(topic, [])
    for n in needles:
        if n and n.lower() in readme_lc:
            return True, n
    return False, None


def main() -> int:
    parser = argparse.ArgumentParser(description="Positioning consistency lint.")
    parser.add_argument("--quiet", action="store_true", help="Suppress success output.")
    args = parser.parse_args()

    h1, anchor, blockquote = _read_readme_anchors()
    pkg_desc = _read_package_description()
    about_desc = _read_about_description()
    topics, equivalents = _read_topics()

    errors: list[str] = []
    anchor_lc = anchor.lower()
    if anchor_lc not in pkg_desc.lower():
        errors.append("package.json.description missing canonical anchor")
    if anchor_lc not in about_desc.lower():
        errors.append(".github/about.yml description missing canonical anchor")
    if len(pkg_desc) > DESCRIPTION_MAX:
        errors.append(f"package.json.description is {len(pkg_desc)} chars (max {DESCRIPTION_MAX})")
    if len(about_desc) > DESCRIPTION_MAX:
        errors.append(f".github/about.yml description is {len(about_desc)} chars (max {DESCRIPTION_MAX})")

    readme_lc = README.read_text(encoding="utf-8").lower()
    missing_topics: list[str] = []
    for topic in topics:
        present, _ = _topic_present(readme_lc, topic, equivalents)
        if not present:
            missing_topics.append(topic)

    if errors or missing_topics:
        sys.stderr.write("❌  positioning drift detected:\n")
        sys.stderr.write(f"        README anchor:         {anchor}\n")
        sys.stderr.write(f"        package.json.desc:     {pkg_desc}\n")
        sys.stderr.write(f"        .github/about.yml:     {about_desc}\n\n")
        for err in errors:
            sys.stderr.write(f"        - {err}\n")
        if missing_topics:
            sys.stderr.write("\n        topics absent from README (literal + equivalents):\n")
            for t in missing_topics:
                sys.stderr.write(f"          - {t}\n")
            sys.stderr.write(
                "\n        Resolve by editing all three to share the canonical anchor,\n"
                "        or extending .github/topics.yml's `equivalents:` map\n"
                "        (or by removing the topic). The README is the canonical phrasing.\n"
            )
        else:
            sys.stderr.write(
                "\n        Resolve by editing all three to share the canonical anchor.\n"
                "        The README is the canonical phrasing; the other two follow it.\n"
            )
        return 1

    if not args.quiet:
        sys.stdout.write(f"✅  positioning consistent (anchor: {anchor!r}, topics: {len(topics)})\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/scripts/lint_positioning.ts
#!/usr/bin/env tsx
/**
 * Positioning consistency lint for event4u/agent-config.
 *
 * TypeScript twin of `src/scripts/lint_positioning.py` (ADR-089, Phase 4 /
 * Wave 4b). Mirrors the CLI contract EXACTLY — the `--quiet` argparse flag,
 * exit codes (0 clean, 1 drift, 2 missing PyYAML — inert here since `yaml`
 * is always present), byte-identical stdout (success line) and stderr
 * (drift report), and the same source files. No behaviour changes.
 *
 * Asserts three public-positioning surfaces agree on the canonical phrasing
 * and that every advertised GitHub topic is discoverable in the README body
 * (literally or through `equivalents:` paraphrases).
 *
 * Sources: README.md (H1 + first blockquote), package.json (`description`),
 * .github/about.yml (`description`), .github/topics.yml (`topics:` +
 * optional `equivalents:` map).
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { parse as parseYaml } from 'yaml';

const _HERE = fileURLToPath(import.meta.url);
const REPO_ROOT = path.resolve(path.dirname(_HERE), '..', '..');
const README = path.join(REPO_ROOT, 'README.md');
const PACKAGE_JSON = path.join(REPO_ROOT, 'package.json');
const ABOUT_YML = path.join(REPO_ROOT, '.github', 'about.yml');
const TOPICS_YML = path.join(REPO_ROOT, '.github', 'topics.yml');

// re.compile(r"^#\s+(.+?)\s*$", re.MULTILINE)
const H1_RE = /^#\s+(.+?)\s*$/m;
// re.compile(r"^>\s+(.+?)\s*$", re.MULTILINE)
const BLOCKQUOTE_RE = /^>\s+(.+?)\s*$/m;
const DESCRIPTION_MAX = 200;

/** Thrown for the SystemExit("…") raises that print and exit 1. */
class SystemExitError extends Error {
    constructor(public readonly text: string) {
        super(text);
        this.name = 'SystemExitError';
    }
}

function _parseYamlSafe(text: string): unknown {
    // version '1.1' matches PyYAML safe_load.
    return parseYaml(text, { version: '1.1' });
}

function _read_readme_anchors(): [string, string, string] {
    const text = fs.readFileSync(README, 'utf-8');
    const h1_match = H1_RE.exec(text);
    if (!h1_match) {
        throw new SystemExitError('❌  README.md has no H1 heading');
    }
    const h1 = h1_match[1] as string;
    // re.split(r"\s+[—–-]\s+", h1, maxsplit=1)
    const parts = _splitMax(h1, /\s+[—–-]\s+/, 1);
    const anchor = parts.length === 2 ? (parts[1] as string).trim() : h1.trim();

    const bq_match = BLOCKQUOTE_RE.exec(text);
    const blockquote = bq_match ? (bq_match[1] as string).trim() : '';
    return [h1, anchor, blockquote];
}

/** re.split(pattern, s, maxsplit=n) — split at most `maxsplit` times. */
function _splitMax(s: string, pattern: RegExp, maxsplit: number): string[] {
    const re = new RegExp(pattern.source, pattern.flags.includes('g') ? pattern.flags : pattern.flags + 'g');
    const out: string[] = [];
    let last = 0;
    let count = 0;
    let m: RegExpExecArray | null;
    while (count < maxsplit && (m = re.exec(s)) !== null) {
        out.push(s.slice(last, m.index));
        last = m.index + m[0].length;
        count += 1;
        if (m.index === re.lastIndex) {
            re.lastIndex++;
        }
    }
    out.push(s.slice(last));
    return out;
}

function _read_package_description(): string {
    const data = JSON.parse(fs.readFileSync(PACKAGE_JSON, 'utf-8'));
    return String(data?.description ?? '').trim();
}

function _read_about_description(): string {
    const data = (_parseYamlSafe(fs.readFileSync(ABOUT_YML, 'utf-8')) ?? {}) as Record<string, unknown>;
    return String(data?.description ?? '').trim();
}

function _read_topics(): [string[], Record<string, string[]>] {
    const data = (_parseYamlSafe(fs.readFileSync(TOPICS_YML, 'utf-8')) ?? {}) as Record<string, unknown>;
    const topicsRaw = Array.isArray(data['topics']) ? (data['topics'] as unknown[]) : [];
    const topics = topicsRaw.map((t) => String(t));
    const equivalentsRaw =
        data['equivalents'] && typeof data['equivalents'] === 'object'
            ? (data['equivalents'] as Record<string, unknown>)
            : {};
    const equivalents: Record<string, string[]> = {};
    for (const [k, vs] of Object.entries(equivalentsRaw)) {
        equivalents[String(k)] = (Array.isArray(vs) ? vs : []).map((v) => String(v));
    }
    return [topics, equivalents];
}

function _topic_present(
    readme_lc: string,
    topic: string,
    equivalents: Record<string, string[]>,
): [boolean, string | null] {
    const needles = [
        topic,
        topic.replace(/-/g, ' '),
        topic.replace(/-/g, ''),
        ...(equivalents[topic] ?? []),
    ];
    for (const n of needles) {
        if (n && readme_lc.includes(n.toLowerCase())) {
            return [true, n];
        }
    }
    return [false, null];
}

interface ParsedArgs {
    quiet: boolean;
}

function _argparse_error(message: string): never {
    process.stderr.write('usage: lint_positioning.py [-h] [--quiet]\n');
    process.stderr.write(`lint_positioning.py: error: ${message}\n`);
    process.exit(2);
}

function parse_args(argv: readonly string[]): ParsedArgs {
    let quiet = false;
    for (const arg of argv) {
        if (arg === '--quiet') {
            quiet = true;
        } else if (arg === '-h' || arg === '--help') {
            process.stdout.write('usage: lint_positioning.py [-h] [--quiet]\n');
            process.exit(0);
        } else {
            _argparse_error(`unrecognized arguments: ${arg}`);
        }
    }
    return { quiet };
}

export function main(): number {
    const args = parse_args(process.argv.slice(2));

    let anchor: string;
    let pkg_desc: string;
    let about_desc: string;
    let topics: string[];
    let equivalents: Record<string, string[]>;
    try {
        [, anchor] = _read_readme_anchors();
        pkg_desc = _read_package_description();
        about_desc = _read_about_description();
        [topics, equivalents] = _read_topics();
    } catch (e) {
        if (e instanceof SystemExitError) {
            process.stdout.write(e.text + '\n');
            return 1;
        }
        throw e;
    }

    const errors: string[] = [];
    const anchor_lc = anchor.toLowerCase();
    if (!pkg_desc.toLowerCase().includes(anchor_lc)) {
        errors.push('package.json.description missing canonical anchor');
    }
    if (!about_desc.toLowerCase().includes(anchor_lc)) {
        errors.push('.github/about.yml description missing canonical anchor');
    }
    if (pkg_desc.length > DESCRIPTION_MAX) {
        errors.push(`package.json.description is ${pkg_desc.length} chars (max ${DESCRIPTION_MAX})`);
    }
    if (about_desc.length > DESCRIPTION_MAX) {
        errors.push(`.github/about.yml description is ${about_desc.length} chars (max ${DESCRIPTION_MAX})`);
    }

    const readme_lc = fs.readFileSync(README, 'utf-8').toLowerCase();
    const missing_topics: string[] = [];
    for (const topic of topics) {
        const [present] = _topic_present(readme_lc, topic, equivalents);
        if (!present) {
            missing_topics.push(topic);
        }
    }

    if (errors.length > 0 || missing_topics.length > 0) {
        process.stderr.write('❌  positioning drift detected:\n');
        process.stderr.write(`        README anchor:         ${anchor}\n`);
        process.stderr.write(`        package.json.desc:     ${pkg_desc}\n`);
        process.stderr.write(`        .github/about.yml:     ${about_desc}\n\n`);
        for (const err of errors) {
            process.stderr.write(`        - ${err}\n`);
        }
        if (missing_topics.length > 0) {
            process.stderr.write('\n        topics absent from README (literal + equivalents):\n');
            for (const t of missing_topics) {
                process.stderr.write(`          - ${t}\n`);
            }
            process.stderr.write(
                '\n        Resolve by editing all three to share the canonical anchor,\n' +
                    "        or extending .github/topics.yml's `equivalents:` map\n" +
                    '        (or by removing the topic). The README is the canonical phrasing.\n',
            );
        } else {
            process.stderr.write(
                '\n        Resolve by editing all three to share the canonical anchor.\n' +
                    '        The README is the canonical phrasing; the other two follow it.\n',
            );
        }
        return 1;
    }

    if (!args.quiet) {
        process.stdout.write(`✅  positioning consistent (anchor: ${_pyRepr(anchor)}, topics: ${topics.length})\n`);
    }
    return 0;
}

/** Python `repr()` of a string literal (single-quote preference). */
function _pyRepr(s: string): string {
    if (s.includes("'") && !s.includes('"')) {
        return `"${s.replace(/\\/g, '\\\\')}"`;
    }
    return `'${s.replace(/\\/g, '\\\\').replace(/'/g, "\\'")}'`;
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    process.exit(main());
}

export {
    REPO_ROOT,
    README,
    PACKAGE_JSON,
    ABOUT_YML,
    TOPICS_YML,
    DESCRIPTION_MAX,
    _read_readme_anchors,
    _topic_present,
};
